File: app/api/src/Astra.Api/Llm/Archetypes/dotnet8/canonical-php-catalog-service/src/CartService.cs
// SPDX-Spec: php/add_to_cart.php, php/cart.php (signed)
// SPDX-Archetype: canonical-php-catalog-service (dotnet8)
namespace Acme.Catalog;

/// <summary>
/// Explicit projection of the PHP <c>$_SESSION['cart']</c> superglobal (SG-1) —
/// ambient session state lifted to an injected boundary (on promotion an
/// <c>ISession</c>-backed scoped service), never ambient static state.
/// </summary>
[SpecClaim("SG-1")]
[TargetMapping("scoped service over ISession", PhpConstruct = "$_SESSION['cart']")]
public interface ISessionCart
{
    /// <summary>Current cart keyed by sku (⟵ $_SESSION['cart']); never null.</summary>
    IDictionary<string, CartLine> GetCart();

    /// <summary>Upsert a line back into the session cart (a side effect, SE-1).</summary>
    [SpecClaim("SE-1")]
    void PutLine(CartLine line);
}

/// <summary>
/// C# projection of PHP add_to_cart.php + cart.php. Reads/writes the cart via
/// <see cref="ISessionCart"/> instead of touching $_SESSION directly; adding is a
/// side effect (SE-1); the total is exact decimal money (INV-1).
/// </summary>
[TargetMapping("service class", PhpConstruct = "add_to_cart.php / cart.php")]
public sealed class CartService
{
    private readonly ISessionCart _session;

    public CartService(ISessionCart session) => _session = session;

    /// <summary><c>$_SESSION['cart'][$sku] = ($existing ?? 0) + $qty</c>.</summary>
    [SpecClaim("SG-1")]
    [SpecClaim("SE-1")]
    public int AddToCart(string sku, int qty, decimal price)
    {
        if (qty < 0)
            throw new ArgumentException("qty must be non-negative", nameof(qty));

        var cart = _session.GetCart();
        var merged = cart.TryGetValue(sku, out var existing)
            ? existing.AddQty(qty)
            : new CartLine(sku, qty, price);
        _session.PutLine(merged); // SE-1: write-back
        return merged.Qty;
    }

    /// <summary><c>cartTotal($cart)</c> — Σ (qty × price), as exact decimal money.</summary>
    [SpecClaim("ARR-1")]
    [SpecClaim("INV-1")]
    public decimal CartTotal()
    {
        decimal total = 0m;
        foreach (var line in _session.GetCart().Values)
            total += MoneyMath.LineTotal(line.Price, line.Qty);
        return Math.Round(total, 2, MidpointRounding.AwayFromZero);
    }
}

File: app/api/src/Astra.Api/Llm/Archetypes/java-spring/canonical-php-catalog-service/src/main/java/com/acme/catalog/CartService.java
// SPDX-Spec: php/add_to_cart.php, php/cart.php (signed)
// SPDX-Archetype: canonical-php-catalog-service
package com.acme.catalog;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.Map;

/**
 * Java 21 projection of PHP {@code add_to_cart.php} + {@code cart.php}. The cart
 * (a PHP associative array, ARR-1) is read from and written back to the session
 * via {@link SessionCartPort} (SG-1), instead of touching {@code $_SESSION}
 * directly. Adding to the cart is a real side effect (SE-1); the total is exact
 * decimal money (INV-1).
 */
@TargetMapping(value = "@Service", phpConstruct = "add_to_cart.php / cart.php")
public final class CartService {

    private final SessionCartPort session;

    public CartService(SessionCartPort session) {
        this.session = session;
    }

    /**
     * {@code $_SESSION['cart'][$sku] = ($existing ?? 0) + $qty} — merge a line
     * into the session cart and return the new quantity for that sku.
     */
    @SpecClaim("SG-1")
    @SpecClaim("SE-1")
    public int addToCart(String sku, int qty, BigDecimal price) {
        if (qty < 0) {
            throw new IllegalArgumentException("qty must be non-negative");
        }
        CartLine existing = session.getCart().get(sku); // ?? 0 → null-check below
        CartLine merged = (existing == null)
            ? new CartLine(sku, qty, price)
            : existing.addQty(qty);
        session.putLine(merged); // SE-1: write-back to the session
        return merged.qty();
    }

    /** {@code cartTotal($cart)} — Σ (qty × price), as exact decimal money. */
    @SpecClaim("ARR-1")
    @SpecClaim("INV-1")
    public BigDecimal cartTotal() {
        BigDecimal total = BigDecimal.ZERO;
        for (Map.Entry<String, CartLine> e : session.getCart().entrySet()) {
            CartLine line = e.getValue();
            total = total.add(MoneyMath.lineTotal(line.price(), line.qty()));
        }
        return total.setScale(2, RoundingMode.HALF_UP);
    }
}
